zkvm/docs/zkvm-spec.md

```diff
@@ -371,6 +374,42 @@ Time bounds are available in the transaction as [expressions](#expression-type)
 [`mintime`](#mintime) and [`maxtime`](#maxtime).
 
 
+### Contract ID
+
+_Contract ID_ is a 32-byte commitment to the [contract](#contract-type) ([anchor](#anchor), [payload](#contract-payload) and [predicate](#predicate)) that is also a unique identifier.
+
+Contract ID is a hash of the contract’s [output structure](#output-structure):
+
+```
+T = Transcript("ZkVM.contractid")
+T.commit("contract", output_structure)
+id = T.challenge_bytes()
+```
+
+Contract ID makes [`delegate`](#delegate) signatures safe against [predicate](#predicate) key reuse:
+signature covers the unique contract ID, therefore preventing its replay against another contract,
+even if containing the same [payload](#contract-payload) and predicate.
+
+Contract ID is made globally unique due to use of [anchor](#anchor).
+
+
+### Anchor
+
+_Anchor_ is a 32-byte unique string that provides uniqueness to the [contract ID](#contract-id).
+
+Anchors can be created by the [`nonce`](#nonce) instruction or generated from previously used unique contract IDs, tracked by the VM via [last anchor](#vm-state):
+
+1. Nonce contract has its anchor computed from the nonce parameters (see [`nonce`](#nonce) instruction).
+2. Claimed UTXO ([`input`](#input)) sets the VM’s [last anchor](#vm-state) to its [contract ID](#contract-id).
+3. Transient contract ([`contract`](#contract)) consumes the VM’s [last anchor](#vm-state) and replaces it with its [contract ID](#contract-id).
```

Comment:
Is it safe for `contract` to use last-anchor as-is? If so, can you say why? Or should it be ratcheted as with `output`?

Reply:
WDYT?

> Transient contract sets its ID as an anchor because this can be done only once: the contract is soon destroyed by `signtx`, `call` or `delegate`.
> This is not the case with `output`: if the output’s ID was set as the anchor, it'd be available in two transactions: in this one, for the rest of the execution, and in the future one, when the output is claimed by `input` which sets the same output ID as an anchor.

Reply:
OK ✔️